### PR TITLE
Add SQLite Grammar

### PR DESCRIPTION
**Pull request by jsteinshouer**
Issue: #219

Here is an attempt at adding support for SQLite. I have a couple items I would like your feedback on:

### Limited support for `ALTER TABLE`

SQLite only supports a limited subset of `ALTER TABLE`.  It **does not** support:

- Modifying columns
- Modifying constraints
- Droping constraints

I currently coded it so these functions with throw a "UnsupportedOperation" exception. The SQLite docs suggest a workaround by creating a new table migrating the data to it then dropping the old table. Let me know if this is something you would like to support and I can try to implement it.

https://www.sqlite.org/lang_altertable.html##otheralter

### No `CONCAT` function

I did some testing with your [up-and-running-with-quick](https://github.com/Ortus-Solutions/up-and-running-with-quick) application and noticed that Quick uses the `CONCAT` function. For some reason SQLite does not have that function but uses `||` to concatenate strings together.

I ended up creating a function named `replaceConcat` which basically intercepts the sql statement before execution and then replaces `CONCAT` with `||`. It seems a bit hacky to me but left it in there. I can remove it if you prefer.

### CODE REVIEW

File: models/Grammars/SQLiteGrammar.cfc [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+    ===================================*/
+
+		/**
+		 * Replace instances of CONCAT function with || since SQLite does not support CONCAT
+		 *
+		 * @sql The sql string.
+		 *
+		 * @return string
+		 */
+		private string function replaceConcat( sql ) {
```
Comment (elpete) [inline models/Grammars/SQLiteGrammar.cfc]: Perhaps we could introduce a `concat` helper method on qb that does the concat for that the grammar supports and then upgrade Quick to use that.

File: models/Grammars/SQLiteGrammar.cfc [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+        /**
+         * Compiles the lock portion of a sql statement.
+         *
+         * @query The Builder instance.
+         * @lockType The lock type to compile.
+         *
+         * @return string
+         */
+        private string function compileLockType( required query, required string lockType ) {
+            return "";
```
Comment (elpete) [inline models/Grammars/SQLiteGrammar.cfc:31]: So does SQLite not support any table locks?

**Review by elpete — APPROVED** [review]

**Comment by elpete** [comment]
Could you include in the README a way to get a SQLite database running and added as a datasource in ACF and Lucee?

File: models/Grammars/SQLiteGrammar.cfc (hunk 1 above)
Comment (elpete) [reply models/Grammars/SQLiteGrammar.cfc]: I added a `concat` helper method in `next` that you'll want to implement for SQLite.

File: models/Grammars/SQLiteGrammar.cfc (hunk 2 above)
Comment (jsteinshouer) [reply models/Grammars/SQLiteGrammar.cfc:31]: I could not find any documentation on table locks. If I am understanding this correctly, the locking is at the database level and transactions are commited all at once.

https://www.sqlite.org/lockingv3.html

File: models/Grammars/SQLiteGrammar.cfc (hunk 1 above)
Comment (jsteinshouer) [reply models/Grammars/SQLiteGrammar.cfc]: Thanks. I will take a look and get that implemented.

Comment (jsteinshouer) [reply models/Grammars/SQLiteGrammar.cfc]: I have implemented the `concat` method and removed the replaceConcat method I was using. I also updated the tests based on recent changes. I will take a look at updating the readme next.

**Comment by jsteinshouer** [comment]
> Could you include in the README a way to get a SQLite database running and added as a datasource in ACF and Lucee?

I updated the README with instructions for setting up the driver and datasource. Let me know if looks good.

**Comment by elpete** [comment]
This is excellent.  Thank you so much!

**Pull request merged**